adds complex tag management

bcoca opened the pull request:
Adds a special tag:
- always: always runs no matter what --tags/--skip-tags says

Adds 4 special keywords for --tags
- all: all tagged + untagged tasks
- tagged: only tagged tasks
- untagged: only untagged tasks
- always: only run tasks tagged 'always'

I think this covers all cases described in various tickets asking for more fine control over tags

bcoca committed:
adds complex tag management

Adds a special tag:
 - always: always runs no matter what --tags, --skip-tags say

Adds 4 special keywords for --tags/skip-tags
 - all: all tagged + untagged tasks
 - tagged: only tagged tasks
 - untagged: only untagged tasks
 - always: only run tasks tagged 'always'

bcoca committed:
fixed skipped tags

Ernest0x commented:
> always: only run tasks tagged 'always'

I think it should be:
- always: always run tasks tagged with "always" whatever the '--tags' command-line option is set (or not set) to

brian-brazil commented:
Taking a quick look based on by work on #7039:
- The tag code generally needs a refactor to correctly handle --list-tasks, as the code is currently duplicated incorrectly
- `always` should also bypass --start-at
-  There should be checks that tasks aren't using reserved tags
-  Unittests should be added

I'd suggest taking my PR, and adding the additional functionality on top of it as it handles the above already.

brian-brazil commented on lib/ansible/playbook/__init__.py:
```diff
@@ -730,19 +735,41 @@ def _run_play(self, play):
+                                should_run = False
+                        else:
+                            if should_run:
+                                if len(task_set.intersection(self.skip_tags)) > 0:
```
You can do:

```
if task_set.intersection(self.skip_tags):
```

bcoca replied:
good point, I'll update

arnuschky commented:
This covers my use case, but the `always` keyword/tag would suffice. 

What does the `all` keyword to over omitting `--tags` altogether?

For me, `tagged` and `untagged` are only marginally interesting: I always tags everything, so `untagged` might only help in debugging.

hkariti commented:
One thing I see missing here that has been requested in the past is the ability to AND multiple tags. Something like: `--tags conf:&app1`

mscherer commented:
So, what would "--skip-tags always" do, and what would --skip-tags "all" do ? 

The second seems like a no-op, but the first is more puzzling. If it skip always, then always is not always. If if doesn't skip anything, then that's a exception to the rule.

brian-brazil commented:
`--skip-tags always` should not run the always tags. There are disaster recovery cases where you may not wish some `always` tasks to run (e.g. some backend you need is down, causing the task to fail), and this is one way to handle that.

bcoca commented:
current state of this request:
- "--tags all" is the actual default if you omit --tags, so it is also a noop when specified (this is true w/o  patch)
- "--skip-tags always" doesn't do anything, things tagged 'always' cannot be skipped
- "--skip-tags all" will only execute tasks with 'always'
- "--tags always" is basically the same as "--skip-tags all"
- --start-at-task is not tag aware, imho it should not be
- --skip-tags overrides --tags (this is also true w/o patch)

I left out tests,list-tasks and task expressions (matching host expressions) as I wanted to nail down the main functionality first. The 2 points of discord I still see are that some don't want 'always' to be always and the inclusion of --start-at-task in tag management.

brian-brazil commented:
> the inclusion of --start-at-task in tag management.

I don't think there's discord here. Only @tkinz27 and I have mentioned this on #7039, and both of us appear to want `always` to override --start-at-task.

bcoca committed:
simplified if conditions

bcoca committed:
added tests
refactored task selection into method
list-tasks now uses common method of task selection
always is now skippable if specified in --skip-tags

bcoca commented:
@brian-brazil copied your tests (I hope you don't mind) and copied your use of a method to reuse task selection in ansible-playbook

I also changed 'always' tag to be skippable

I'm not adding star-at-task functionality nor tag pattern matching in this round, leaving those for future PRs

bcoca committed:
documented new tag behaviour

brian-brazil commented:
I'm happy to get this feature in, so please take whatever is useful from my PR. 

You may want to pull in the deletion of `_is_valid_tag` in lib/ansible/playbook/play.py, as that's dead code that would cause confusion given the new code.

bcoca committed:
removed unused method

sevein commented:
Nice!
